### PR TITLE
components: Add isValueNumeric util

### PR DESCRIPTION
**Pull request by sarayourfriend**
## Description
Adds a new `isValueNumeric` utility required by `TextInput`.

Does this need to also take in a locale?

## How has this been tested?
Unit tests pass

## Types of changes
New feature

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [N/A] My code follows the accessibility standards. 
- [N/A] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

**Comment by aristath** [comment]
Is `isValueNumeric` different to [`_.isNumber()`](https://lodash.com/docs/4.17.15#isNumber)?

**Comment by sarayourfriend** [comment]
`isValueNumeric` handles comma separated values. But I think it needs to take in a locale to know whether to strip commas or periods 🤷 

We could use `_.isNumber` internally to `isValueNumeric` though, after stripping the delimiter, rather than handling that stuff ourselves.

Edit: We can't use `isNumber` because `isNumber('4') === false`. We need to be able to parse strings here.

**Comment by sarayourfriend** [comment]
I'm going to convert this to a hook that uses `useI18n` to retrieve the locale data and returns an `isNumericValue` function.

**Comment by aristath** [comment]
Thank you for taking the time to explain this, it makes sense now :+1:

**Comment by sarayourfriend** [comment]
Actually I don't think it's possible to get the delimiter from the locale data, not even using `Itnl.NumberFormat`. I think this is the best we can do.

**Comment by sarayourfriend** [comment]
@yuliyan Do you mind taking a look a this approach to localization for this utility? I think I have something workable but just want to make sure from an i18n/l10n perspective that it makes sense. Thanks!

**Comment by sarayourfriend** [comment]
Hmmmm, that failing unit test passes locally. I'm not sure how to fix that issue. I'm going to try debugging it tomorrow by pushing some debugging commits with print statements and such.

**Comment by ciampo** [comment]
> @yuliyan Do you mind taking a look a this approach to localization for this utility? I think I have something workable but just want to make sure from an i18n/l10n perspective that it makes sense. Thanks!

Tagging also @dlind1 for some additional feedback

**Comment by sarayourfriend** [comment]
Summarizing a discussion about the unit tests that we had in the #core-js channel today: https://wordpress.slack.com/archives/C5UNMSU4R/p1626189861190800

JavaScript unit tests are run against Node versions 12 and 14.
The test fails in version 12 because version 12 doesn’t ship with full-icu data for the internationalization stuff whereas 14 does. The default changed for Node between versions 12 and 14 (see documentation links below).

https://nodejs.org/docs/latest-v12.x/api/intl.html#intl_options_for_building_node_js
https://nodejs.org/docs/latest-v14.x/api/intl.html#intl_options_for_building_node_js

We still support Node 12 until EOL which is in 2022.

https://github.com/jsdom/jsdom/issues/1626

Three options (from @nerrad):
a. Drop support for Node 12 (breaking with what we're following with other libraries)
b. Don't use Intl.NumberFormat yet?
c. Do the checks only have this warn in non browser environments?
d. Require Node 12 built with `--full-icu`

Option b seems wild to me in that we’d be limiting our usage of native browser APIs in a browser app based on non-browser constraints

Option c would be something like re-writing isValueNumeric to be naïve about localization in non-browser environments (i.e.,`typeof document === 'undefined’`)? But that doesn’t solve the fact that we want to unit test the Intl.NumberFormat code too right? Because to get the tests to pass we’d probably end up changing the test/values.js to run in a non-jsdom environment… Otherwise the behavior would stay the same.

Option a which seems like a big departure from the community for the sake of a small utility function.

Option d is a cumbersome requirement for users of our libraries. If we went with that though we could enforce it by doing something like the following:

```js
function hasFullIcu() {
  return new Intl.NumberFormat('pt-BR').format(1000.1) === '1.000,1';
}
```

The magical Option 🤩 that @gziolo came up with was to just scope the unittest to run in Node 14.

I'll go ahead and go with Option 🤩 🙂

File: packages/components/src/utils/test/values.js [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+ * @see https://nodejs.org/docs/latest-v12.x/api/intl.html#intl_options_for_building_node_js
+ * @see https://nodejs.org/docs/latest-v14.x/api/intl.html#intl_options_for_building_node_js
+ * @see https://github.com/jsdom/jsdom/issues/1626
+ *
+ * @todo Remove when Node 12 is deprecated
+ * @param {() => void} testCallback
+ */
+function scopeTestToFullICU( testCallback ) {
+	if ( Intl.NumberFormat.supportedLocalesOf( 'pt-BR' ).length === 1 ) {
+		testCallback();
```
Comment (gziolo) [inline packages/components/src/utils/test/values.js:25]: That's an interesting trick with a callback 👍🏻 

In other places, we wrap `describe` with a similar logic, but the end result is exactly the same.

**Comment by gziolo** [comment]
> The magical Option 🤩 that @gziolo came up with was to just scope the unittest to run in Node 14.
> 
> I'll go ahead and go with Option 🤩 🙂

The only inconvenience would be when someone uses Node 12 for development as this test doesn't run, but it isn't very unlikely and also not recommended in the contributors' guide.

Besides, every test that runs conditionally is tricky but in this case it's a legitimate expception.

**Comment by yuliyan** [comment]
@sarayourfriend, that's very interesting approach. I did some testing and it seems to be working pretty well for the most part.

One thing that I noticed is that the delimiter check for big numbers is a bit loose, e.g. `isValueNumeric( '2,2,2,2,2,2,2.22', 'hi-IN' );` would return `true`, though I'm not sure if that should be considered a problem.

Also, it would be good to consider handling spaces as delimiter in addition to the locale's specific delimiter. AFAIK, it's the internationally recommended thousands separator - https://en.wikipedia.org/wiki/Decimal_separator.

**Comment by sarayourfriend** [comment]
> Also, it would be good to consider handling spaces as delimiter in addition to the locale's specific delimiter. AFAIK, it's the internationally recommended thousands separator

Wow! I had no idea, I'll add logic to support that as well.

Thanks for your review @yuliyan 🙇

**Comment by sarayourfriend** [comment]
> One thing that I noticed is that the delimiter check for big numbers is a bit loose, e.g. isValueNumeric( '2,2,2,2,2,2,2.22', 'hi-IN' ); would return true, though I'm not sure if that should be considered a problem.

That's interesting, I hadn't considered that the _format_ might be invalid.

To solve that we could always compare against the return value of `Intl.NumberFormat.prototype.format`ing the final parsed value against the input value to see if they're the same... But again that doesn't seem like an issue like you said and might be _too_ strict for this utility.

File: packages/components/src/utils/test/values.js [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+	}
+}
+
+describe( 'isValueNumeric', () => {
+	scopeTestToFullICU( () => {
+		it( 'should handle space separated numbers for various locales', () => {
+			expect( isValueNumeric( '1 000.1', 'en-US' ) ).toBe( true );
```
Comment (ciampo) [inline packages/components/src/utils/test/values.js]: Tiny nit, but — Could this specific line be run outside of the `scopeTestToFullICU()` wrapper?

**Comment by ciampo** [comment]
> might be too strict for this utility.

I think this depends on what's the goal of this function, because technically `2,2,2,2,2,2,2.22` is _not_ a valid numeric value. If this function is going to be used to, e.g., validate numeric input fields, then I would argue that we should be more strict.

> We could always compare against the return value of `Intl.NumberFormat.prototype.formating` the final parsed value against the input value to see if they're the same

This seems like a reasonable approach to me, similar in principle to how the separators are computed too. Another alternative could be to make sure that each `separator` is followed by 3 digits in the input string.

**Comment by sarayourfriend** [comment]
I really strongly do not believe the function should validate the format of the number. It's too high complexity for this function. If we want to enforce format like that it should be a separate utility meant for that and that alone.

**Comment by ciampo** [comment]
> I really strongly do not believe the function should validate the format of the number. It's too high complexity for this function. If we want to enforce format like that it should be a separate utility meant for that and that alone.

That's also a valid point. Maybe we should add in the JSDoc of `isValueNumeric` that we're purposefully not checking for a number to have a correct format.

We could then implement a separate utility if needed, as suggested above

**Review by ciampo — APPROVED** [review]
LGTM 🚀

**Comment by dlind1** [comment]
Could this have some issues with Arabic locales? `Intl.NumberFormat( 'ar' ).format( 1000.1 );` returns  a number with Arabic decimal dot and comma `١٬٠٠٠٫١`, so it wouldn't remove the English dot/comma if the entered number used them.

**Comment by sarayourfriend** [comment]
> Could this have some issues with Arabic locales? Intl.NumberFormat( 'ar' ).format( 1000.1 ); returns a number with Arabic decimal dot and comma ١٬٠٠٠٫١, so it wouldn't remove the English dot/comma if the entered number used them.

Ah, it does seem like that would cause a problem.

I wonder what the best approach is in this case. Any advice @dlind1 ?

**Comment by sarayourfriend** [comment]
I guess we could say, if the decimal or thousands delimiter is not a comma or period AND the number itself is using ~latin~ Western Arabic numerals (I guess a regex should suffice for that) then to just use the `'en-US'` locale? That seems pretty naïve... I'm not really sure what a good solution for this is.

**Comment by dlind1** [comment]
I'd expect that if the number is entered using Arabic or Persian numerals, then all the delimiters would also be localized. So we might have two scenarios:
1. The locale is 'ar', 'fa' etc. and the number is localized - we could use a regex to detect it, then do something like number.replace(/[٠-٩]/g, d => '٠١٢٣٤٥٦٧٨٩'.indexOf(d)).replace(/[۰-۹]/g, d => '۰۱۲۳۴۵۶۷۸۹'.indexOf(d)).replace(/٬/g,',').replace(/٫/g,'.');` to get an English number of decimal point separator, after which we could use the 'en-US' locale.
2. The number is not using Arabic/Persian numerals, but the locale is 'ar', 'fa' etc. - just use the 'en-US' locale (or maybe 'en-GB', depends on what is the convention for entering numbers using Arabic-Western numerals).

File: packages/components/src/utils/values.js [hunk 3]
```diff
@@ -40,3 +40,60 @@ export function isValueEmpty( value ) {
...
+	return [
+		new RegExp( `\\${ delimiter }`, 'g' ),
+		new RegExp( `\\${ decimal }`, 'g' ),
+	];
+};
+
+// https://en.wikipedia.org/wiki/Decimal_separator#Current_standards
+const INTERNATIONAL_THOUSANDS_DELIMITER = / /g;
+
+const ARABIC_NUMERAL_LOCALES = [ 'ar', 'fa', 'ur' ];
```
Comment (dlind1) [inline packages/components/src/utils/values.js]: Could ckb (Central Kurdish) and ps (Pashto) be added as well?

File: packages/components/src/utils/values.js [hunk 4]
```diff
@@ -40,3 +40,60 @@ export function isValueEmpty( value ) {
...
+ * @param {string} [locale]
+ * @return {boolean} Whether value is numeric.
+ */
+export function isValueNumeric( value, locale = window.navigator.language ) {
+	if ( ARABIC_NUMERAL_LOCALES.some( ( l ) => locale.startsWith( l ) ) ) {
+		locale = 'en-GB';
+		if ( EASTERN_ARABIC_NUMBERS.test( value ) ) {
+			value = value
+				.replace( /([۰-۹]|[٠-٩])/g, ( /** @type {string} */ d ) =>
+					'٠١٢٣٤٥٦٧٨٩'.indexOf( d )
```
Comment (dlind1) [inline packages/components/src/utils/values.js:81]: The numbers are slightly different for ar/ckb and fa/ps. You could see them by running `Intl.NumberFormat( 'ar' ).format( 123456789.01 );` vs `Intl.NumberFormat( 'fa' ).format( 123456789.01 );`.

Comment (sarayourfriend) [reply packages/components/src/utils/values.js:81]: Ah, I see that now, thanks.

**Review by dlind1 — APPROVED** [review]
LGTM!

**Pull request merged**